### PR TITLE
feat: Add FP8 quantization and dequantization support with scaling for torch-spyre

### PR DESCRIPTION
**Pull request by Akash-Nayak**
## What type of PR is this?

* [ ] bug
* [x] feature
* [ ] documentation
* [ ] cleanup

## What this PR does

This PR adds FP8 quantization support to torch-spyre.

### Changes

* Adds `quantize_fp8_with_scale` operation for FP8 quantization.
* Implements FP8 quantization using the `qfp8ch` backend operation.
* Adds support for FP8 formats
  
* Integrates FP8 quantization with `torch.compile(backend="inductor")`.
* Uses hardware-accelerated `reciprocal`, `multiply`, `clamp`, and quantization operations.

### Implementation Notes
During validation of `qfp8ch`, a synthetic-dimension workaround was used to verify FP16→FP8 conversion correctness and isolate the known "anywhere invalid" issue (https://github.com/torch-spyre/torch-spyre/issues/2185).

The workaround:

1. Adds a temporary synthetic dimension before quantization.
2. Executes FP8 quantization through qfp8ch.
3. Removes the synthetic dimension after quantization.

## Which issue(s) this PR is related to

https://github.com/torch-spyre/torch-spyre/issues/1538
https://github.com/torch-spyre/torch-spyre/issues/1538?issue=torch-spyre%7Ctorch-spyre%7C1542
https://github.com/torch-spyre/torch-spyre/issues/2185
https://github.com/torch-spyre/torch-spyre/issues/2252

#### Usage

```python
x_fp8 = torch.ops.spyre.quantize_fp8_with_scale(x, scale)
```

Where:

- x is an FP16 input tensor on SPYRE.
- scale is an FP16 scaling factor used during quantization.
- The returned tensor is an FP8 tensor (torch.float8_e4m3fn).
- Validation was performed using a synthetic-dimension workaround to isolate the "anywhere invalid" issue.

## Does this PR introduce a user-facing change?

Yes.

This PR introduces FP8 quantization support in torch-spyre, enabling reduced memory footprint and improved efficiency for large-model inference workloads.

### Validation

- Verified FP16 → FP8 conversion on SPYRE using `torch.compile(backend="inductor")`.
- Validated output correctness against expected FP8 quantization behavior.

### CODE REVIEW

**Comment by avery-blanchard** [comment]
@Akash-Nayak could you fix the commit signatures?

**Comment by Akash-Nayak** [comment]
> @Akash-Nayak could you fix the commit signatures?

Sure @avery-blanchard , I have fixed the commit signatures. Thanks!

File: torch_spyre/_inductor/customops.py [hunk 1]
```diff
@@ -458,3 +458,98 @@ def to_dtype_cpu(input: torch.Tensor, dtype: torch.dtype) -> torch.Tensor:
 @to_dtype_cpu.register_fake
 def _(input: torch.Tensor, dtype: torch.dtype) -> torch.Tensor:
     return torch.empty_like(input, dtype=dtype)
+
+
+@torch.library.custom_op("spyre::reciprocal", mutates_args=(), device_types="spyre")
+def reciprocal(input: torch.Tensor) -> torch.Tensor:
```
Comment (avery-blanchard) [inline torch_spyre/_inductor/customops.py]: Is there a reason we need a custom operation / lowering for reciprocal here?

Comment (Akash-Nayak) [reply torch_spyre/_inductor/customops.py]: Good catch! I have now removed the custom reciprocal operation and switched to `lowering.reciprocal()` instead.

**Comment by avery-blanchard** [comment]
Could we add tests for the roundtrip conversion here?
copy fp16 tensor to device->`quantize_fp8_with_scale`->`fp8_to_dl16`->copy fp16 tensor from device

**Comment by Akash-Nayak** [comment]
> Could we add tests for the roundtrip conversion here? copy fp16 tensor to device->`quantize_fp8_with_scale`->`fp8_to_dl16`->copy fp16 tensor from device

Sure, I have added comprehensive tests for the quantization and dequantization custom ops in commit 77bd162. The tests cover:

- Roundtrip conversion: FP16 → FP8 (quantize) → FP16 (dequantize)
- Multiple tensor shapes and scale values

While running these tests, I encountered two bugs related to symbolic expressions when using dynamic shapes:

1. Layout propagation bug - TypeError when evaluating symbolic modulo expressions in boolean context
2. Scratchpad allocator bug - TypeError when converting symbolic coefficients to integers

I have created PR #2511 to fix both of these bugs. The Inductor Ops FP8 (pull_request) test in this PR should pass once PR #2511 gets merged.

File: torch_spyre/_inductor/spyre_kernel.py [hunk 2]
```diff
@@ -427,8 +431,12 @@ def create_op_spec(
                 op in [IDENTITY_OP, RESTICKIFY_OP]
                 or DtypeOpTable.is_dtype_op(op)
                 or (op in SPYRE_FP32_OPS and arg.device_dtype == DataFormats.IEEE_FP32)
-                or arg.device_dtype == DataFormats.SEN169_FP16
-                or arg.device_dtype == DataFormats.SEN143_FP8
+                or arg.device_dtype
+                in [
+                    DataFormats.SEN169_FP16,
+                    DataFormats.SEN143_FP8,
+                    DataFormats.SEN152_FP8,
```
Comment (yoheiueda) [inline torch_spyre/_inductor/spyre_kernel.py]: Do all ops support these types? If some don’t, it might be safer to add an op-specific check here to avoid unintended behavior.

Comment (Akash-Nayak) [reply torch_spyre/_inductor/spyre_kernel.py]: Good point. I have added the `SPYRE_FP8_OPS` whitelist and an op-specific check so that FP8 dtypes are allowed only for verified FP8-supported ops. Any unsupported op/FP8 combination will raise an Unsupported error, preventing unintended behavior.

File: tests/inductor/test_inductor_ops.py [hunk 3]
```diff
@@ -4095,6 +4100,156 @@ def setUp(self):
                 "4d_dim3": (3, cached_randn((2, 4, 8, 64))),
             },
         },
+        ("test_dequantize_fp8_with_scale", "test_dequantize_fp8_with_scale_cpu"): {
```
Comment (avery-blanchard) [inline tests/inductor/test_inductor_ops.py]: Can we use `cached_rand` here and use the same atol/rtol as the tests in the other fp8 test file?
Also, is there a reason why the tests are split across two files? Is it alright if we move them all to `test_fp8_operations.py`?

File: tests/inductor/test_inductor_ops_lx_planning.py [hunk 4]
```diff
@@ -243,6 +243,7 @@ def new_test(self, value=value):
     "test_unbind_3d_dimneg1",
     "test_unbind_4d_dim0",
     "test_unbind_4d_dim3",
+    "test_dequantize_fp8_with_scale_shape_1_2_32_scale_0.01_mean_0.0_std_1.0",
```
Comment (avery-blanchard) [inline tests/inductor/test_inductor_ops_lx_planning.py]: This change should not longer be required after the fp8 tests were moved out of `tests/inductor/test_inductor_ops.py`

File: tests/inductor/test_inductor_ops.py [hunk 5]
```diff
@@ -289,7 +289,7 @@ def _dtype_name(dt):
     )
     for src, dst in DtypeOpTable.get_dtype_pairs()
     for shape in TO_DTYPE_OP_SHAPES
-    if src != torch.bool and dst != torch.bool
+    if src not in (torch.bool, torch.float8_e4m3fn) and dst != torch.bool
```
Comment (avery-blanchard) [inline tests/inductor/test_inductor_ops.py:292]: Is this change still required after moving the fp8 tests?

File: tests/inductor/test_fp8_operations.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        For scale=2.0: input ~600 / 2.0 = ~300 (FP8 spacing 32), error * 2.0 = 64.0
+        Example: 336.0 quantizes to either 320.0 or 352.0, giving 32 error, scaled to 64.
+        """
+        # Generate deterministic input tensor with specified mean and std
+        # cached_randn only supports scale parameter, so we apply mean/std transformation manually
+        x = cached_randn(shape, dtype=torch.float16, scale=1.0) * std + mean
+        scale = torch.tensor([scale_value], dtype=torch.float16)
+
+        # Calculate atol based on scale: max FP8 spacing * scale * safety factor
+        atol = FP8_E4M3_MAX_SPACING * scale_value * TOLERANCE_SAFETY_FACTOR
```
Comment (avery-blanchard) [inline tests/inductor/test_fp8_operations.py]: I think that the atol here is too relaxed for most of the inputs tested here. 
I understand the extreme cases for larger input ranges where this could be necessary, but maybe those should be split off into a separate test?

File: tests/inductor/test_inductor_ops.py (hunk 3 above)
Comment (Akash-Nayak) [reply tests/inductor/test_inductor_ops.py]: Updated the tests to use `cached_randn` and consolidated all FP8-related tests into test_fp8_operations.py.

I also split the quantize/dequantize tests into three categories:

`test_quantize_dequantize_fp8_small_range`
test_quantize_dequantize_fp8_medium_range`
`test_quantize_dequantize_fp8_boundary_cases`

The production-shape tests continue to use the same tolerance (atol=0.5, rtol=0.1) as the existing FP8 conversion tests. For the quantization/dequantization tests, larger tolerances are required for medium and boundary cases because FP8 E4M3 quantization errors can be amplified during dequantization, especially near representation boundaries.

File: tests/inductor/test_inductor_ops.py (hunk 5 above)
Comment (Akash-Nayak) [reply tests/inductor/test_inductor_ops.py:292]: I kept the torch.float8_e4m3fn exclusion in test_inductor_ops.py.

This exclusion is independent of the FP8 quantization tests. It prevents generic dtype conversion tests from attempting conversions involving torch.float8_e4m3fn, which are not currently supported by the underlying test path.

File: tests/inductor/test_fp8_operations.py (hunk 6 above)
Comment (Akash-Nayak) [reply tests/inductor/test_fp8_operations.py]: Agreed. I split the tests into separate categories based on the expected FP8 quantization error:

Small range (mean=0, std={1,5}) → atol=2.0
Medium range (mean=10, std=50) → atol=16.0
Boundary cases (mean/std={100,100} and {200,200}) → atol=32 * scale

This allows the majority of test cases to use tighter tolerances while reserving the larger, scale-dependent tolerance only for inputs that intentionally exercise FP8 E4M3 representation boundaries.

File: tests/inductor/test_inductor_ops_lx_planning.py (hunk 4 above)
Comment (Akash-Nayak) [reply tests/inductor/test_inductor_ops_lx_planning.py]: Removed obsolete FP8 test reference from `test_inductor_ops_lx_planning.py`
Deleted: `test_dequantize_fp8_with_scale_shape_1_2_32_scale_0.01_mean_0.0_std_1.0`. This test no longer exists after moving FP8 tests to test_fp8_operations.py

**Review by tardieu — DISMISSED** [review]
lgtm

**Comment by tardieu** [comment]
@ashokponkumar please review the test matrix changes.

**Comment by tardieu** [comment]
@JRosenkranz @thoangtrvn could you please review the cpp changes?

**Review by ashokponkumar — DISMISSED** [review]

File: tests/inductor/test_fp8_operations.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
+# Copyright 2024 IBM Corp.
```
Comment (thoangtrvn) [inline tests/inductor/test_fp8_operations.py]: wrong or old license header

File: torch_spyre/_inductor/decompositions.py [hunk 8]
```diff
@@ -792,6 +811,16 @@ def where_scalar_decomp(condition, self, other):
     return torch.ops.aten.where.self(condition, self_t, other_t)
 
 
+@register_spyre_decomposition([torch.ops.spyre.quantize_fp8_with_scale])
+def spyre_quantize_fp8_with_scale(
+    input: torch.Tensor, scale: torch.Tensor
+) -> torch.Tensor:
+    inv_scale = torch.reciprocal(scale)
+    x_scaled = input * inv_scale
+    x_clamped = torch.ops.spyre.clamp(x_scaled, -448.0, 448.0)
```
Comment (thoangtrvn) [inline torch_spyre/_inductor/decompositions.py]: can we use named constant for this magic number? e.g. `FP8_E4M3_MAX`  - we can put it in `_inductor/constants.py` file

File: tests/inductor/test_fp8_operations.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        - Roundtrip: FP16 → FP8 → FP16 with scaling
+        - Verifies qfp8ch operation is used internally
+
+        Note: We use dequantize_fp8_with_scale for FP8→FP16 conversion
+        because direct .to(torch.float16) cannot transfer to CPU.
+        """
+        x = cached_randn((1, 2, 8), scale=1.0, dtype=torch.float16)
+        scale = torch.ones((1, 2, 1), dtype=torch.float16)
+
+        def spyre_fn(x, scale):
```
Comment (thoangtrvn) [inline tests/inductor/test_fp8_operations.py:54]: qfp8ch is only validated via a quant→dequant round-trip with loose tolerances, i.e. against a PyTorch reference at atol=0.5, rtol=0.1. Two issues:

1. The single assertion folds two device ops together, so a qfp8ch error and a fp8todl16 error can compensate for each other and still pass.
2. The tolerance is wide enough to hide a 1-ULP quantization error landing on the wrong E4M3 data point.
qfp8ch is pure E4M3 format conversion (deterministic round-to-nearest-even), so it should be testable at atol=0. 
We also want to check that value beyond +/-448 must clamp.

**Review by thoangtrvn — COMMENTED** [review]
I have a few comments.

**Review by tardieu — APPROVED** [review]
lgtm

**Comment by tardieu** [comment]
@ashokponkumar @thoangtrvn this PR needs another round of approvals.

**Review by ashokponkumar — APPROVED** [review]

File: tests/inductor/test_fp8_operations.py [hunk 10]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2024 IBM Corp.
+# Copyright 2025 The Torch-Spyre Authors.
```
Comment (thoangtrvn) [inline tests/inductor/test_fp8_operations.py:1]: should it be 2026?

**Review by thoangtrvn — APPROVED** [review]
LGTM
